### ISSUE TITLE
VIO rate drops sporadically

### ISSUE THREAD

**Issue by plusk01** [mit-acl/snap-vislam-pkgs#1]
There are two versions of VISLAM ROSified by ATLFlight:
- [`ros-examples`](https://github.com/ATLFlight/ros-examples): Single node which handles camera, imu, cpa, and vislam and only broadcasts `/vislam/pose` and `/vislam/odometry` (no images)
- [`snap_vio`](https://github.com/ATLFlight/snap_vio) (which this repo submodules): nodelet-based implementation (also need `snap_cam_ros`, `snap_imu`, `snap_msgs`, `snap_cpa`) and broadcasts `/downward/image_raw`, `/downward/vio/pose`, etc.

**The concern**: When the vio package (whichever) and `imu_app -s 2` are running, I get 30 hz vio. When all the other acl flight packages are running, vio rate drops (sometimes to 20-25 hz, sometimes to 10-15, etc) and is sporadic (sometimes it even becomes difficult to use the snapdragon computer).

At one point, I was seeing that `ros-examples` was getting better performance than `snap_vio`. Better in the sense that `rostopic hz /HX02/vislam/pose -w60` was much more constant than `rostopic hz /downward/vio/pose -w60`.

At one point, my hypothesis was a power issue (the PDB board output 5V/2A, but Intrinsyc recommends 5V/4A). But I saw the same behaviour when plugged into the wall (5V/4A). *A good test would be to use the current sensor in the APM to measure how much the snapdragon is pulling.*

Thoughts:
- I have successfully flown with vio at 30 hz using `snap_vio standalone.launch`. I was not running the `relay` python script. I recorded data offboard (compressed) and onboard (only images).
- After a `catkin_make` things seemed to go south
- `while true; do true; done` doesn't seem to really affect it
- When `hz` starts getting sporadic, I notice in `top` that the nodelet goes from it's nominal 20% to 30-40% (Irix off). Could be a ROS thing?

Seems like a classic case of a heisenbug

![image](https://user-images.githubusercontent.com/3934603/62083043-0f7e1a00-b224-11e9-9dd3-ca2a2b80012b.png)
https://xkcd.com/1780/

**Comment by plusk01** [comment]
It seems like there is a correlation with temperature

**Comment by plusk01** [comment]
temperature relationship confirmed: https://developer.qualcomm.com/comment/12090#comment-12090

It is still not clear that this is the only affect I was seeing.

**Comment by plusk01** [comment]
I believe the temperature-based throttling is the root cause of the sporadic output rate issue (the `snap_cpa` nodelet sometimes not working is another issue).

Intrinsyc does sell fans as part of the [stereovision kit](https://www.intrinsyc.com/qualcomm-flight/)

![image](https://user-images.githubusercontent.com/3934603/62473214-4d23fb00-b76e-11e9-87db-708c5410700f.png)

There is also [this](https://www.adafruit.com/product/3368) Adafruit Miniature 5v cooling fan that we could buy and design/3d print the fan mount ourselves. However, I've noticed that propwash from the hex is enough to keep the snapdragon cool. I started flying when the temperature was ~87 and it quickly dropped down to 50-60. So the issue I was chasing was because I was on the ground and is really a non-issue for flight.
